### PR TITLE
[FLINK-15124][table] Fix types with precision defined in DDL can't be executed

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Fix the problem that types with precision defined in DDL can't be executed.
The reason is that when code generation, the result type of field access expression is derived from `TableSource#getProducedDataType` (which may lose precision), and the expected result type is from DDL (which keeps precision). And comparing these two types, an exception is thrown. 

## Brief change log

- Add a `DataTypePrecisionFixer` to fix precision with corresponding logical schema type
- Fixes precision before generate code for converting source output to internal row
- Update CsvTableSource to use DataType instead of TypeInformation.

## Verifying this change

- added a IT casee to read a table with precision types and write into a table with precision types.
- added a IT casee to verify CSV with precision types 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
cc @JingsongLi

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java [hunk 1]
```diff
@@ -75,6 +75,8 @@
 		properties.add(SCHEMA + ".#." + DescriptorProperties.TABLE_SCHEMA_TYPE);
 		properties.add(SCHEMA + ".#." + DescriptorProperties.TABLE_SCHEMA_DATA_TYPE);
 		properties.add(SCHEMA + ".#." + DescriptorProperties.TABLE_SCHEMA_NAME);
+		// schema watermark
+		properties.add(SCHEMA + "." + DescriptorProperties.WATERMARK + ".*");
```
Comment (JingsongLi) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: Why need watermark?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sources/CsvTableSourceFactoryBase.java [hunk 2]
```diff
@@ -85,6 +85,8 @@
 		properties.add(SCHEMA + ".#." + DescriptorProperties.TABLE_SCHEMA_TYPE);
 		properties.add(SCHEMA + ".#." + DescriptorProperties.TABLE_SCHEMA_DATA_TYPE);
 		properties.add(SCHEMA + ".#." + DescriptorProperties.TABLE_SCHEMA_NAME);
+		// schema watermark
+		properties.add(SCHEMA + "." + DescriptorProperties.WATERMARK + ".*");
```
Comment (JingsongLi) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sources/CsvTableSourceFactoryBase.java]: Same question to sink.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala [hunk 3]
```diff
@@ -223,10 +224,10 @@ object SinkCodeGenerator {
           case (fieldTypeInfo, i) =>
             val requestedTypeInfo = tt.getTypeAt(i)
             validateFieldType(requestedTypeInfo)
-            if (!areTypesCompatible(
-              fromTypeInfoToLogicalType(fieldTypeInfo),
-              fromTypeInfoToLogicalType(requestedTypeInfo)) &&
-              !requestedTypeInfo.isInstanceOf[GenericTypeInfo[Object]]) {
+            if (!PlannerTypeUtils.isAssignable(
+                fromTypeInfoToLogicalType(fieldTypeInfo),
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala]: checkstyle?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala [hunk 4]
```diff
@@ -100,18 +96,32 @@ object TableSourceUtil {
...
-        val (physicalName, idx, tpe) = resolveInputField(name, tableSource)
+        val (physicalName, idx, logicalType) = resolveInputField(name, tableSource)
         // validate that mapped fields are are same type
-        if (!isAssignable(fromTypeInfoToLogicalType(tpe), t)) {
+        if (!isAssignable(logicalType, t)) {
           throw new ValidationException(s"Type $t of table field '$name' does not " +
-            s"match with type $tpe of the field '$physicalName' of the TableSource return type.")
+            s"match with type $logicalType of the field '$physicalName' of the " +
+            "TableSource return type.")
+        } else if (!isInteroperable(logicalType, t)) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: Why check decimal precision?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala [hunk 5]
```diff
@@ -100,18 +96,32 @@ object TableSourceUtil {
           throw new ValidationException(s"Rowtime field '$name' has invalid type $t. " +
             s"Rowtime attributes must be of TimestampType.")
         }
-        val (physicalName, idx, tpe) = resolveInputField(name, tableSource)
+        val (physicalName, idx, logicalType) = resolveInputField(name, tableSource)
         // validate that mapped fields are are same type
-        if (!isAssignable(fromTypeInfoToLogicalType(tpe), t)) {
+        if (!isAssignable(logicalType, t)) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: If we need isInteroperable, so we can remove this `isAssignable`?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala (hunk 4 above)
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: Source sink maybe can unify? Use same check?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/catalog/CatalogTableITCase.scala [hunk 6]
```diff
@@ -174,6 +179,37 @@ class CatalogTableITCase(isStreamingMode: Boolean) {
...
+        |  price DECIMAL(10, 2),
+        |  currency STRING,
+        |  ts TIMESTAMP(6)
+        |) with (
+        |  'connector' = 'COLLECTION'
+        |)
+      """.stripMargin
+    tableEnv.sqlUpdate(sinkDDL)
+
+    val query = "INSERT INTO T2 SELECT * FROM T1 "
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/catalog/CatalogTableITCase.scala]: test insert into csv too?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala [hunk 7]
```diff
@@ -18,31 +18,27 @@
...
+import org.apache.flink.table.runtime.types.{LogicalTypeDataTypeConverter, TypeInfoLogicalTypeConverter}
+import org.apache.flink.table.runtime.types.PlannerTypeUtils.{isAssignable, isInteroperable}
+import org.apache.flink.table.runtime.types.LogicalTypeDataTypeConverter.fromDataTypeToLogicalType
+import org.apache.flink.table.runtime.types.TypeInfoLogicalTypeConverter.fromLogicalTypeToTypeInfo
 import org.apache.flink.table.sources.{DefinedFieldMapping, DefinedProctimeAttribute, DefinedRowtimeAttributes, RowtimeAttributeDescriptor, TableSource}
 import org.apache.flink.table.types.DataType
-import org.apache.flink.table.types.logical.{LogicalType, TimestampKind, TimestampType, TinyIntType}
+import org.apache.flink.table.types.logical.{LogicalType, LogicalTypeRoot, RowType, TimestampKind, TimestampType, TinyIntType}
 import org.apache.flink.table.typeutils.TimeIndicatorTypeInfo
-
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: unnecessary change

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala [hunk 8]
```diff
@@ -223,10 +224,10 @@ object SinkCodeGenerator {
...
+                fromTypeInfoToLogicalType(requestedTypeInfo)) &&
+                !requestedTypeInfo.isInstanceOf[GenericTypeInfo[Object]]) {
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala]: Should the type be assignable or just equals strictly ? I think it is the later.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java (hunk 1 above)
Comment (wuchong) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: The factory can't be found if a watermark is defined in DDL. 
I will update the IT case to use watermark.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sources/CsvTableSourceFactoryBase.java (hunk 2 above)
Comment (wuchong) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sources/CsvTableSourceFactoryBase.java]: A table defined by DDL can be a source or a sink. So it is also needed for sinks.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala (hunk 4 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: As we discussed offline, if the produced data type of TableSource doesn't carry the correct precision, a readable exception should be thrown. This is a guide to suggest users to update their connectors.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala (hunk 5 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: No, we can't remove it. The intention here is separating type mis-match and precision mis-match.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala (hunk 4 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: I though about this. But I think we can loosen the restriction for sinks, because 
1) The conversion from internal type (Deciaml) to external type (BigDecimal) doesn't loose precision, because the internal type is already match to the expected type defined in DDL.
2) it's pretty hard to give a correct consumed DataType for `UpsertStreamTableSink`.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/catalog/CatalogTableITCase.scala (hunk 6 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/catalog/CatalogTableITCase.scala]: Sure. I can update it.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala (hunk 7 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: Good catch!

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala (hunk 8 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala]: 1. It's safe to be assignable (ignore precision and nullablitiy).
2. Currently, almost all the sinks are using TypeInformations (e.g. UpsertTableSink can only use TypeInformations). And TypeInformations will lose precision informations which will lead to be always type-incompatible.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala (hunk 3 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala]: I added an additional indent to make the code readable (separate from the following code block). But I can revert it if you think the original style is correct.

**Comment by wuchong** [comment]
Hi @JingsongLi , @danny0405 , I have updated the PR, please have another look.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: So every connector(SourceTableFactory and SinkTableFactory) must has this?
Kafka not work too?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala (hunk 3 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala]: I think Scala only one indent is enough. Correct me if I wrong.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala (hunk 4 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: I think you should add comments to both source and sink to explain why they are different.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala (hunk 8 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala]: It can be discussed to whether it should be replaced by `isInteroperable`.
But never `just equals strictly`, it is too strict.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala [hunk 9]
```diff
@@ -100,18 +96,32 @@ object TableSourceUtil {
...
+        if (!isAssignable(logicalType, t)) {
           throw new ValidationException(s"Type $t of table field '$name' does not " +
-            s"match with type $tpe of the field '$physicalName' of the TableSource return type.")
+            s"match with type $logicalType of the field '$physicalName' of the " +
+            "TableSource return type.")
+        } else if (!isInteroperable(logicalType, t)) {
+          // the produced type of TableSource is different with the logical type defined in DDL
+          // on the precision or nullability.
+          throw new ValidationException(
+            "If the connector would like to support precision and nullability defined in DDL," +
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: No `nullability` check, both `isInteroperable` and `isAssignable` not check `nullability`.

Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala]: OK. I will update the exception message and comment.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java (hunk 1 above)
Comment (wuchong) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: Yes. I will fix it in following PR with some computed column problems.

Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: Can you create a JIRA to track it too? Should in 1.10?

Comment (wuchong) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: issue created FLINK-15191.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala (hunk 3 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala]: OK.

**Comment by wuchong** [comment]
@JingsongLi , after thinking again, I think the planner should take care of the precision instead of pushing the work to connector developers. 
I updated the PR to fix the precisison before code generation, please have a another look.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixer.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+import org.apache.flink.table.types.logical.ZonedTimestampType;
+
+import java.util.Map;
+
+/**
+ * The data type visitor used to fix the precision for data type with the given logical type
+ * which carries the correct precisions. The original data type may loses precision because
+ * of conversion from {@link org.apache.flink.api.common.typeinfo.TypeInformation}.
+ */
+public final class DataTypePrecisionFixer implements DataTypeVisitor<DataType> {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixer.java]: I am a little worried about scan all precision types.
Another way is just bridge all conversion types to DataType from DDL schema, what do you think?

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixer.java]: Unfortunately we can't, in some special cases, the original data type is not the same to logical type, e.g. BIGINT vs Timestamp Rowtime Attribute, and Legacy(Pojo) vs RowType. So currently, we only picks the precision types to update.

**Comment by JingsongLi** [comment]
Code LGTM, but I think you can have some tests manually.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: side comment: at some point we need to refactor this such that not every connector much list all properties. This is also why we introduced `TableSourceFactory#createTableSource(org.apache.flink.table.catalog.ObjectPath, org.apache.flink.table.catalog.CatalogTable)`
The outer layer should deal with the schema properties and the connector can work with `CatalogTable.getTableSchema` if necessary.

Comment (wuchong) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/CsvTableSinkFactoryBase.java]: Exactly, agree.

**Comment by wuchong** [comment]
I verified in my local machine and it works well in SQL CLI. 
Will merge it.
